### DISCUSSION TITLE
Merge Queue: allow multiple merge methods

### DISCUSSION THREAD

**Discussion by lukaspiatkowski** [community/community#17162] [Pull Requests]
This is about [Merge Queue](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request-with-a-merge-queue), the currently closed beta feature of GitHub.

**Problem statement**

The Merge Queue allows one of the 3 methods for merging: "Merge commit", "Squash and merge", "Rebase and merge". It is possible only to choose one of them.

**Desired behavior**

Merge Queue should allow multiple merging methods to allow different workflows of many PR authors in a project, e.g. some might want to simply merge their commit while others might want to sqash it before merging. The merge method is choosable in a non-queued version of merge, so it should be the same for Merge Queue.

**Comment by matthiasbeyer** [comment]
Let me add to that: AFAICT merge-queues does an octopus-merge when it is merging "normally" (no rebase involved). There should be an option to resort to one-branch-per-merge instead of having octopus merge and potentially ten or even more merges in one merge!

**Comment by sandstrom** [comment]
For the merge queue to work for our org, we'd want to configure it such that all (a) PRs are rebased on top of master before being merged, and that (b) a merge commit is always used for the merge.

**Comment by torressam333** [comment]
Just ran into this discussion with our Architect. 
in 95% of the scenarios using the "Squash and Merge" option out of the box from merge queues would suffice BUT what about the other 5% of the time where we don't want to squash and merge?

Is there a way to allow the developer to choose the Merge Method at the time of Adding their PR to the queue?
Also, is it possible to have the merge queue do this:

1. Allow selecting merge method via UI
2. Have merge queue Rebase the PR automagically against the target branch (i.e., dev) when added to the queue?

**Reply by matthiasbeyer** [reply]
> n 95% of the scenarios using the "Squash and Merge" option out of the box from merge queues would suffice BUT what about the other 5% of the time where we don't want to squash and merge?

FWIW, I **never** want to squash and merge, because squash effectively means deleting history!

> Have merge queue Rebase the PR automagically against the target branch (i.e., dev) when added to the queue?

If this is an option, fine. But I wouldn't use it because it breaks signatures (as in GPG/SSH signed commits).

**Reply by sandstrom** [reply]
I think you're effectively asking for this feature:

https://github.com/orgs/community/discussions/14863

If so, feel free to upvote it.

**Reply by pfac** [reply]
> because squash effectively means deleting history!

Well, you're not wrong, but that is not necessarily deleting, nor a bad thing.

Let's say you were working on something in your own branch. Before opening the PR you squash all your "wip" commits into a single one, with a proper message and everything tidy. Throughout the PR you're asked to do a couple of changes, small things that don't change the feature. Squashing those "code review" commits into the first one with that proper message is a perfectly valid use case. I would personally prefer that to having "wip" messages in the main branch's history.

But to each their own 🙂. If we could have the option for you not to use squash, and everyone who wants to be able to, that I would say is the perfect outcome.

**Reply by matthiasbeyer** [reply]
Maybe I didn't phrase this out 100% correctly: I did not mean that I am against squashing in general, but against squashing a PR branch just because of squashing. A PR branch that consists of more than one commit but with a nice history does not need squashing.

Also, cleaning up WIP commits (which IMO are completely fine during the development and review cycle of a PR, and I often myself have PRs that are more `--fixup` commits than "actual" commits) is the task of the _requester_, not of the maintainer of a project.

**Reply by torressam333** [reply]
We squash and merge for dev but do merge commits to preserve history on master. We want dev history clean. We only occasionally need to Merge Commit on dev if for example we rename a file...we don't want history for the old file name to be lost once we change extensions or name etc...otherwise we want to keep history clean on dev.
